src/libraries/System.Text.Json/tests/System.Text.Json.SourceGeneration.Tests/UnspeakableTypeTests.cs

```diff
@@ -0,0 +1,146 @@
+
+            // Serializing using the declared type works as expected
+            string json = await Serializer.SerializeWrapper(envelope.Value, options);
+            Assert.Equal(expectedJson, json);
+
+            // But it can also be serialized as an object
+            json = await Serializer.SerializeWrapper<object>(envelope.Value, options);
+            Assert.Equal(expectedJson, json);
+
+            // or if you pass in its runtime type 
+            json = await Serializer.SerializeWrapper(envelope.Value, envelope.Value.GetType(), options);
+            Assert.Equal(expectedJson, json);
+
+            // And it can be deserialized using the declared type
+            await Serializer.DeserializeWrapper<T>(json, options);
+
+            // But will still fail if you attempt to deserialize using the runtime type
+            await Assert.ThrowsAsync<NotSupportedException>(() => Serializer.DeserializeWrapper(json, envelope.Value.GetType(), options));
+        }
+
+        [Theory]
+        [MemberData(nameof(GetUnspeakableTypes))]
+        public async Task CanSerializeUnspeakableTypesAsBoxedProperties<T>(Envelope<T> envelope, string expectedJson)
+        {
+            var boxedEnvelope = new Envelope<object>(envelope.Value);
+            string expectedEnvelopeJson = $$"""{"Value":{{expectedJson}}}""";
+
+            string json = await Serializer.SerializeWrapper(boxedEnvelope, UnspeakableTypeContext.Default.Options);
+            Assert.Equal(expectedEnvelopeJson, json);
+        }
+
+        [Theory]
+        [MemberData(nameof(GetUnspeakableTypes))]
+        public async Task CanSerializeUnspeakableTypesAsBoxedCollectionElements<T>(Envelope<T> envelope, string expectedJson)
+        {
+            var boxedEnvelope = new List<object> { envelope.Value };
+            string expectedCollectionJson = $"[{expectedJson}]";
+
+            string json = await Serializer.SerializeWrapper(boxedEnvelope, UnspeakableTypeContext.Default.Options);
+            Assert.Equal(expectedCollectionJson, json);
+        }
+
+        [Theory]
+        [MemberData(nameof(GetUnspeakableTypes))]
+        public async Task CanSerializeUnspeakableTypesAsBoxedDictionaryValues<T>(Envelope<T> envelope, string expectedJson)
+        {
+            var boxedEnvelope = new Dictionary<string, object> { ["key"] = envelope.Value };
+            string expectedCollectionJson = $$"""{"key":{{expectedJson}}}""";
+
+            string json = await Serializer.SerializeWrapper(boxedEnvelope, UnspeakableTypeContext.Default.Options);
+            Assert.Equal(expectedCollectionJson, json);
+        }
+
+        [Fact]
+        public async Task TypeWithDiamondAmbiguityThrowsNotSupportedException()
+        {
+            object value = new TypeWithDiamondAmbiguity();
+            await Assert.ThrowsAsync<NotSupportedException>(() => Serializer.SerializeWrapper(value, UnspeakableTypeContext.Default.Options));
```

Comment:
What's the exception message in the case of a diamond ambiguity? Should it be different than if there were no candidates at all?

Reply:
Good point, I've updated to include information on the ambiguity.